parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveComparator.java

```diff
@@ -0,0 +1,202 @@
+    }
+  }
+
+  static final PrimitiveComparator<Integer> SIGNED_INT32_COMPARATOR = new IntComparator() {
+    @Override
+    public int compare(int i1, int i2) {
+      return Integer.compare(i1, i2);
+    }
+  };
+
+  static final PrimitiveComparator<Integer> UNSIGNED_INT32_COMPARATOR = new IntComparator() {
+    @Override
+    public int compare(int i1, int i2) {
+      // Implemented based on com.google.common.primitives.UnsignedInts.compare(int, int)
+      return Integer.compare(i1 ^ Integer.MIN_VALUE, i2 ^ Integer.MIN_VALUE);
+    }
+  };
+
+  private static abstract class LongComparator extends PrimitiveComparator<Long> {
+    @Override
+    public final int compare(Long o1, Long o2) {
+      return compare(o1.longValue(), o2.longValue());
+    }
+  }
+
+  static final PrimitiveComparator<Long> SIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      return Long.compare(l1, l2);
+    }
+  };
+
+  static final PrimitiveComparator<Long> UNSIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      // Implemented based on com.google.common.primitives.UnsignedLongs.compare(long, long)
+      return Long.compare(l1 ^ Long.MIN_VALUE, l2 ^ Long.MIN_VALUE);
+    }
+  };
+
+  static final PrimitiveComparator<Float> FLOAT_COMPARATOR = new PrimitiveComparator<Float>() {
+    @Override
+    public int compare(Float o1, Float o2) {
+      return compare(o1.floatValue(), o2.floatValue());
+    }
+
+    @Override
+    public int compare(float f1, float f2) {
+      return Float.compare(f1, f2);
+    }
+  };
+
+  static final PrimitiveComparator<Double> DOUBLE_COMPARATOR = new PrimitiveComparator<Double>() {
+    @Override
+    public int compare(Double o1, Double o2) {
+      return compare(o1.doubleValue(), o2.doubleValue());
+    }
+
+    @Override
+    public int compare(double d1, double d2) {
+      return Double.compare(d1, d2);
+    }
+  };
+
+  private static abstract class BinaryComparator extends PrimitiveComparator<Binary> {
+    @Override
+    public final int compare(Binary o1, Binary o2) {
+      return compare(o1.toByteBuffer(), o2.toByteBuffer());
+    }
+
+    abstract int compare(ByteBuffer b1, ByteBuffer b2);
+
+    final int toUnsigned(byte b) {
+      return b & 0xFF;
+    }
+  }
+
+  static final PrimitiveComparator<Binary> LEXICOGRAPHICAL_BINARY_COMPARATOR = new BinaryComparator() {
+    @Override
+    int compare(ByteBuffer b1, ByteBuffer b2) {
+      int l1 = b1.remaining();
+      int l2 = b2.remaining();
+      int p1 = b1.position();
+      int p2 = b2.position();
+      int minL = Math.min(l1, l2);
+
+      for (int i = 0; i < minL; ++i) {
+        int result = unsignedCompare(b1.get(p1 + i), b2.get(p2 + i));
+        if (result != 0) {
+          return result;
+        }
+      }
+
+      return l1 - l2;
+    }
+
+    private int unsignedCompare(byte b1, byte b2) {
+      return toUnsigned(b1) - toUnsigned(b2);
```

Comment:
This shouldn't use `-` to implement comparison because Java integers wrap around instead of throwing an overflow exception. Instead, use something like this (from [Comparators gist](https://gist.github.com/rdblue/7e80e9700d508d395b004c4c146dbfe2)):
```
        int cmp = Integer.compare(
            ((int) buf1.get(b1pos + i)) & 0xff,
            ((int) buf2.get(b2pos + i)) & 0xff);
```

Reply:
Here we are dealing with byte values casted to ints. There is no chance for overflows.